lib/ip_blocklist.js

```diff
@@ -0,0 +1,165 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * This module contains all the functionality needed to initialize, load
+ * and check an ip blocklist. The ip blocklist file must contain
+ * at least one column that has ip addresses or an ip address range
+ * in the following formats, 192.168.0.0 or 192.168.0.0/32. The blocklist
+ * polls for any file changes and reloads itself if needed.
+ *
+ */
+var Promise = require('bluebird')
+var readFile = Promise.promisify(require('fs').readFile)
+var statFile = Promise.promisify(require('fs').stat)
+var parse = Promise.promisify(require('csv-parse'))
+var path = require('path')
+var ip = require('ip')
+
+module.exports = function (log, config) {
+
+  function getBaseIp(addr, prefixLength) {
+    return ip.mask(addr, ip.fromPrefixLen(prefixLength))
+  }
+
+  function parseRow(ipColumn) {
+    // Ignore # lines
+    if (ipColumn.length > 0 && ipColumn[0] === '#') {
+      return
+    }
+
+    // Parse row for ip address components, supports 1.10.16.0/20 or 1.10.16.0.
+    // Converts them to a base IP address and a prefix length.
+    var tokens = ipColumn.split('/')
+    if (tokens.length === 2) {
+      var prefixLength = parseInt(tokens[1], 10)
+      return {
+        baseIp: getBaseIp(tokens[0], prefixLength),
+        prefixLength: prefixLength
+      }
+    }
+
+    return {
+      baseIp: tokens[0],
+      prefixLength: 32
+    }
+  }
+
+  function IPBlocklist() {
+    this.prefixLengths = []
+    this.ipsByPrefixLength = {}
+    this.fileLastModified = 0
+    this.pollInterval = null
+  }
+
+  IPBlocklist.prototype.load = function (filePath) {
+    var self = this
+
+    // Resolve file path to an absolute location
+    filePath = path.resolve(filePath)
+    self.filePath = filePath
+    self.fileName = path.basename(filePath)
+
+    var startTime = Date.now()
+    return statFile(filePath)
+      .then(function (fileStats) {
+        self.fileLastModified = fileStats.mtime
+        self.fileSize = fileStats.size
+        return readFile(filePath, 'utf8')
+      })
+      .then(function (data) {
+        return parse(data)
+      })
+      .then(function (rows) {
+        rows.forEach(function (row) {
```

Comment:
> (or better yet, some abstracted-away looping equivalent)

I thought bluebird had such a thing, but I can't see it in their API documentation.  One way to hack it might be something like:

```
Promise.each(rows, function(row, idx) {
  ...do stuff with the row...
  if (idx % 1000 === 0) {
    return P.fromNode(setImmediate)
  }
})
```